### PR TITLE
No trailing . in hs-source-dirs entries

### PR DESCRIPTION
**Pull request by shayne-fletcher**
Fixes issue https://github.com/digital-asset/ghc-lib/issues/256.

### CODE REVIEW

File: ghc-lib-gen/src/Ghclibgen.hs [hunk 1]
```diff
@@ -100,7 +100,7 @@ ghcLibParserHsSrcDirs forParserDepends ghcFlavor lib =
         -- 'GhcVersion.hs' IIRC).
         [ stage0GhcBoot | ghcFlavor `elem` [ GhcMaster, Ghc8101, Ghc8102, Ghc901 ] ] ++
         map takeDirectory cabalFileLibraries ++
-        askFiles lib "hs-source-dirs:"
+        filter (not . isSuffixOf ".") (askFiles lib "hs-source-dirs:")
```
Comment (hurryabit) [inline ghc-lib-gen/src/Ghclibgen.hs:103]: I'm not convinced this is the right approach. For instance, if we only have `foo/.` in the list but not `foo`, then there will be no `foo` left afterward. I think doing what was suggested in the issue and normalizing file paths seems more appropriate. [`System.FilePath.normalise`](https://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:normalise) could be very helpful for this.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs:103]: I looked into this very deeply. I won't bother you with the details of why but the fix is correct (but am happy to do so if you want an explanation). That said I have no problems with throwing a `normalise` in the mix so stand by for that.

Comment (shayne-fletcher) [reply ghc-lib-gen/src/Ghclibgen.hs]: I agree with you that using `normalise` is more robust and therefore better for future proofing.

**Comment by shayne-fletcher** [comment]
Ready for another look when you are @hurryabit . Thank-you!

**Comment by shayne-fletcher** [comment]
Before (this is on ghc-master; the trailing dot problem only entered recently after the API reorganization):
```
    hs-source-dirs:
        ghc-lib/stage0/libraries/ghc-boot/build
        ghc-lib/stage0/compiler/build
        libraries/template-haskell
        libraries/ghc-boot-th
        libraries/ghc-boot
        libraries/ghc-heap
        libraries/ghci
        compiler
        compiler/.
```
After:
```
    hs-source-dirs:
        ghc-lib/stage0/libraries/ghc-boot/build
        ghc-lib/stage0/compiler/build
        libraries/template-haskell
        libraries/ghc-boot-th
        libraries/ghc-boot
        libraries/ghc-heap
        libraries/ghci
        compiler
```

**Review by hurryabit — APPROVED** [review]
This looks much better. Thanks for fixing it!

**Comment by shayne-fletcher** [comment]
I think the last commit will deal with the Windows build failure. May I have another `/azp run` please?

**Pull request merged**